HowTo on creating a backend

anokfireball opened the pull request:
Working to resolve #35.

This walks a developer through the process of creating the skeleton for a backend that handles JSON-files (no actual JSON-handling will be implemented).

Internal structures will be refactored if development can be streamlined or simplified.

anokfireball committed:
Replace map with parameter object

In preparation for the HOWTO on backends, streamline the workflow:
- Replace Attribute maps with type-safe and intuitive objects in backend
- Supply a recommended interface for future backend implementations

anokfireball committed:
Begin HowTo on backend

Follow the backend creation with JSON as an example.
Includes File Formats and IOHandlers.

ax3l commented on docs/source/dev/backend.rst:
```diff
@@ -1,6 +1,132 @@
 .. _development-backend:
 
-How to Write a Backend
+How to write a Backend
```
In English titles, all words but for/to/a/... start uppercase :)

anokfireball committed:
Complete the main part of the HowTo

Includes IOTasks and Task Queue

anokfireball committed:
Docstrings for AbstractIOHandlerImpl

ax3l commented on CMakeLists.txt:
```diff
@@ -17,9 +17,9 @@ set(CMAKE_CXX_STANDARD_REQUIRED True)
-set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
-set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
-set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
+set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_SOURCE_DIR}/lib)
```
the prefix `${CMAKE_BINARY_DIR}` must not be changed, just the path behind it is adjustable.

temporary build directories are in the build dir, not in the source (which should stay untouched during build :) )

anokfireball committed:
More docstrings for AbstractIOHandlerImpl

anokfireball committed:
Include API examples in CTEST

Revert CMAKE_*_OUTPUT_DIRECTORY to in-build rather than in-source.
Enable CI test of all API examples if samples are available.
Fix filename extension in test cases.

Commit:
Merge branch 'dev' into topic/howto_backend

anokfireball committed:
Disable compression on HDF5 datasets

Ignore HDF5 dataset compression and transformation for the moment to
sidestep further backend problems.

ax3l commented on CMakeLists.txt:
```diff
@@ -507,21 +507,30 @@ endif()
+    foreach(examplename ${openPMD_EXAMPLE_NAMES})
+        if(${examplename} MATCHES "^.*_parallel$")
+            if(openPMD_HAVE_MPI)
+                add_test(NAME MPI.${examplename}
```
Maybe we want to add `Example.` as a prefix for the name here?

ax3l commented on CMakeLists.txt:
```diff
@@ -507,21 +507,30 @@ endif()
+                        )
+            endif()
+        else()
+            add_test(NAME Serial.${examplename}
```
Maybe we want to add `Example.` as a prefix for the name here?

ax3l commented on docs/source/dev/backend.rst:
```diff
@@ -3,4 +3,269 @@
+merely offers a small set of elementary IO operations.
+
+On the very basic level, you will need to derive a class from ``AbstractIOHandler``:
+
```
I you like, you can include the interface of `AbstractIOHandler` here as follows:

```rst
.. doxygenclass:: openPMD::AbstractIOHandler
   :project: openPMD
   :members:
   :undoc-members:
```

ax3l commented on include/openPMD/IO/AbstractIOHandler.hpp:
```diff
@@ -74,7 +74,7 @@ class AbstractIOHandler
 public:
 #if openPMD_HAVE_MPI
-    /** Construct an appropriate specific IOHandler for the desired IO mode.
+    /** Construct an appropriate specific IOHandler for the desired IO mode that may be MPI-aware.
```
typo: might be?
